ui/plugin-system/src/model/visual-editing.ts

```diff
@@ -0,0 +1,35 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { JsonObject } from '@perses-dev/core';
+import React from 'react';
+
+/**
+ * A component for visual editing of a plugin's options.
+ */
+export type OptionsEditor<Options extends JsonObject = JsonObject> = React.ComponentType<OptionsEditorProps<Options>>;
+
+/**
+ * Common props passed to options editor components.
+ */
+export interface OptionsEditorProps<Options extends JsonObject = JsonObject> {
```

Comment:
I'm pretty Typescript-new, but is it something like this:

`Options` is a generic type that will depend on the panel itself, and then this interface somehow guarantees that we use the same `Options` between panel and panel options editor?

Reply:
Yup, you got it!

Reply:
If you're curious, it's the `Options` parameter here on `PanelPlugin` (that's then used on all its props) that ensures for a given plugin, they use the same options type for all the pieces:

https://github.com/perses/perses/pull/528/files#diff-0598e07a5c7c0745e67a4438e870edef8e13758c88f3ae2a0b111453c5d06041R21-R25

Reply:
Ah cool